plugins/csp-timings/src/Plugin.hpp

```diff
@@ -0,0 +1,40 @@
+////////////////////////////////////////////////////////////////////////////////////////////////////
+//                               This file is part of CosmoScout VR                               //
+//      and may be used under the terms of the MIT license. See the LICENSE file for details.     //
+//                        Copyright: (c) 2019 German Aerospace Center (DLR)                       //
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#ifndef CSP_TIMINGS_PLUGIN_HPP
+#define CSP_TIMINGS_PLUGIN_HPP
+
+#include "../../../src/cs-core/PluginBase.hpp"
+#include "../../../src/cs-gui/GuiItem.hpp"
+#include "../../../src/cs-utils/FrameTimings.hpp"
+
+#include <fstream>
+#include <list>
+
+namespace csp::timings {
+
+/// A plugin which uses the built-in timer queries of CosmoScout VR to draw on-screen live frame
+/// timing statistics. This plugin can also be used to export recorded time series to a CSV file.
+class Plugin : public cs::core::PluginBase {
+ public:
+  void init() override;
+  void deInit() override;
+  void update() override;
+
+ private:
+  /// This store the statistics GUI element.
+  std::unique_ptr<cs::gui::GuiItem> mGuiItem;
+
+  bool mEnableRecording  = false;
+  bool mEnableStatistics = false;
+  std::list<std::unordered_map<std::string, cs::utils::FrameTimings::QueryResult>> mRecordedTimings;
```

Comment:
What is the reasoning behind using a `std::list`? As it looks to me, a `std::vector` would be better, since we append only once a frame, but loop through twice a frame.

Reply:
We only `emplace_back()` once a frame during recording. This should be as fast as possible in order to reduce the impact on the timings. Then (after the recording) we loop through the list to retrieve the data. So I think `std::list` is a reasonable choice!

Reply:
I know we already merged, but my mind couldn't let go 🙈. `std::list` should still perform worse, since it has to allocate memory every time you insert a node, while `std::vector` only needs to allocate memory if the underlying array is full. Memory allocations are very very expensive. I did a small benchmark, here are the results:

std::list: 
```c++
void list() {
    spdlog::info("list");
    spdlog::stopwatch sw;
    std::list<std::array<uint64_t, 10>> l;

    for (size_t i = 0; i < ELEMENTS; ++i) {
        std::array<uint64_t, 10> a{};

        for (auto&& e : a) {
            e = i;
        }

        l.emplace_back(a);
    }

    spdlog::info("  insert: {}", sw);

    uint64_t sum = 0;
    for (auto const& e : l) {
        sum += std::accumulate(e.begin(), e.end(), 0ull);
    }

    spdlog::debug("    Sum: {}", sum);
    spdlog::info("complete: {}\n", sw);
}
```

std::vector:
```c++
void vector() {
    spdlog::info("vector");
    spdlog::stopwatch sw;
    std::vector<std::array<uint64_t, 10>> l;

    for (size_t i = 0; i < ELEMENTS; ++i) {
        std::array<uint64_t, 10> a{};

        for (auto&& e : a) {
            e = i;
        }

        l.emplace_back(a);
    }

    spdlog::info("  insert: {}", sw);

    uint64_t sum = 0;
    for (auto const& e : l) {
        sum += std::accumulate(e.begin(), e.end(), 0ull);
    }

    spdlog::debug("    Sum: {}", sum);
    spdlog::info("complete: {}\n", sw);
}
```

Results with 10000000 elements:
```
list
  insert: 5.9572255
complete: 8.2400265

vector
  insert: 3.3486662
complete: 5.0138386
```

I think `std::vector` should be preferred here.

Reply:
While I agree that this is an interesting question and that you are maybe right - I hope that you agree that this is definitely a [case of premature optimization](https://stackify.com/premature-optimization-evil/)! Therefore I won't spent much time here...

What happens if you replace the `emplace_back(a)` with a `emplace_back(std::move(a))` (or `push_back(std::move(a))`, this shouldn't make any difference)? `std::arrays` are completely stack-allocated, aren't they? So moving them shouldn't make any difference - you could try using `std::unordered_map` as we are using in the actual code they should be faster to move as most of the memory is heap-allocated (I think).

Reply:
I agree, that we shouldn't waste time over a single line like this, but I think it might be worth to get a better understanding in the long term. The [Core Guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rsl-vector) say that `std::vector` should nearly always be preferred. So I would say it is a premature optimization to use `std::list` at all.

`std::vector<uint64_t, 10>` has actually the same size as `std::unordered_map`, which is 80 bytes.

I think we can leave it as is, since the performance impact here is close to none.